README
genext2fs generates an ext2 filesystem as a normal (non-root) user.
It does not require you to mount the image file to copy files on it,
nor does it require that you become the superuser to make device nodes.

http://sourceforge.net/projects/genext2fs

Please use the genext2fs-devel@lists.sourceforge.net mailing list to
contact the developers and/or report bugs.

See the inluded INSTALL file for instructions on installing genext2fs.


NEWS
This is the 1.4.2 release:

	- fixed: Endless loop with --verbose for large images

Please report bugs to https://github.com/bestouff/genext2fs/issues


test.sh
#!/bin/sh

# This script generates a variety of filesystems and checks that they
# are identical to ones that are known to be mountable, and pass fsck
# and various other sanity checks.

# Passing these tests is preferable to passing test-mount.sh because
# this script doesn't require root, and because passing these tests
# guarantees byte-for-byte agreement with other builds, ports,
# architectures, times of day etc.

set -e

. ./test-gen.lib

md5cmp () {
	md5=`calc_digest`
	if [ x$md5 = x$1 ] ; then
		echo PASS
	else
		echo FAIL
		exit 1
	fi
}

dtest () {
	expected_digest=$1
	shift
	dgen $@
	md5cmp $expected_digest
	gen_cleanup
}

ftest () {
	expected_digest=$1
	shift
	fgen $@
	md5cmp $expected_digest
	gen_cleanup
}

ltest () {
	expected_digest=$1
	shift
	lgen $@
	md5cmp $expected_digest
	gen_cleanup
}

# NB: always use test-mount.sh to regenerate these digests, that is,
# replace the following lines with the output of
# sudo sh test-mount.sh|grep test

dtest 4a25d1109fa03f8519e89a5fc365580f 4096 1024 0
dtest a2c5c3b014510456320373e72a7d4b3f 2048 2048 0
dtest 4af8e2915523207f780e22ff1de29633 1024 4096 0
dtest 537400c0b8f868895ced27af64dc703b 8193 1024 0
dtest f6f6a6124104c8ff5ca2e3cfd8b3f396 8194 1024 0
dtest 63f60f06d4a4858404a09d071b688a7d 8193 4096 0
dtest 851d7cac136b44e1bf461fef27180a1b 8194 2048 0
dtest 518b75cd6651ae864babf3b12a70866b 4096 1024 1
dtest 87b13fb5e0f31e2f13faea02f829730b 1024 4096 1
dtest 548ce3a3bf7d57e350a11c0c274b7795 4096 1024 12288
dtest d6041295c924de32af49ad35a3b37c0d 4096 1024 274432
dtest 2dcd1c07084e616433b43043c1309cc6 9000 1024 8388608
dtest 227d38c05d5860dc039812eed603f20f 4500 2048 8388608
dtest 627c09439a5ed6194900f12f5591d28e 2250 4096 8388608
dtest 84dbb9949b3c1c9d7f3237d3cdaa86b5 20000 1024 16777216
dtest a4ab80a62c0fd09a3be023c77e0307b1 10000 2048 16777216
ftest 9108433a817035cb5306e8217d5e634b 4096 default device_table.txt
ltest dd1a554d9253c84f9d3fff1f4669b4c2 200 1024 123456789
ltest 8c9f6f0c9c15b42e32bef60af88ac6ff 200 1024 1234567890
ltest 9d8be8d71a7743b8c675c83f50d6eff0 200 4096 12345678901


list.h
#ifndef __LIST_H__
#define __LIST_H__

#if STDC_HEADERS
# include <stdlib.h>
# include <stddef.h>
#else
# if HAVE_STDLIB_H
#  include <stdlib.h>
# endif
# if HAVE_STDDEF_H
#  include <stddef.h>
# endif
#endif

#ifndef offsetof
#define offsetof(st, m) \
     ((size_t) ( (char *)&((st *)(0))->m - (char *)0 ))
#endif

#define container_of(ptr, type, member) ({ \
                const typeof( ((type *)0)->member ) *__mptr = (ptr); \
                (type *)( (char *)__mptr - offsetof(type,member) );})

typedef struct list_elem
{
    struct list_elem *next;
    struct list_elem *prev;
} list_elem;

static inline void list_init(list_elem *list)
{
    list->next = list;
    list->prev = list;
}

static inline void list_add_after(list_elem *pos, list_elem *elem)
{
    elem->next = pos->next;
    elem->prev = pos;
    pos->next->prev = elem;
    pos->next = elem;
}

static inline void list_add_before(list_elem *pos, list_elem *elem)
{
    elem->prev = pos->prev;
    elem->next = pos;
    pos->prev->next = elem;
    pos->prev = elem;
}

static inline void list_del(list_elem *elem)
{
    elem->next->prev = elem->prev;
    elem->prev->next = elem->next;
}

static inline void list_item_init(list_elem *elem)
{
    elem->next = elem;
    elem->prev = elem;
}

static inline int list_empty(list_elem *elem)
{
    return elem->next == elem;
}

#define list_for_each_elem(list, curr)            \
    for ((curr) = (list)->next; (curr) != (list); (curr) = (curr)->next)

#define list_for_each_elem_safe(list, curr, next)    \
    for ((curr) = (list)->next, (next) = (curr)->next;    \
         (curr) != (list);                    \
         (curr) = (next), (next) = (curr)->next)

#endif /* __LIST_H__ */


cache.h
#ifndef __CACHE_H__
#define __CACHE_H__

#include "list.h"

#define CACHE_LISTS 256

typedef struct
{
    list_elem link;
    list_elem lru_link;
} cache_link;

typedef struct
{
    /* LRU list holds unused items */
    unsigned int lru_entries;
    list_elem lru_list;
    unsigned int max_free_entries;

    unsigned int entries;
    list_elem lists[CACHE_LISTS];
    unsigned int (*elem_val)(cache_link *elem);
    void (*freed)(cache_link *elem);
} listcache;

static inline void
cache_add(listcache *c, cache_link *elem)
{
    unsigned int hash = c->elem_val(elem) % CACHE_LISTS;
    int delcount = c->lru_entries - c->max_free_entries;

    if (delcount > 0) {
        /* Delete some unused items. */
        list_elem *lru, *next;
        cache_link *l;
        list_for_each_elem_safe(&c->lru_list, lru, next) {
            l = container_of(lru, cache_link, lru_link);
            list_del(lru);
            list_del(&l->link);
            c->entries--;
            c->lru_entries--;
            c->freed(l);
            delcount--;
            if (delcount <= 0)
                break;
        }
    }

    c->entries++;
    list_item_init(&elem->lru_link); /* Mark it not in the LRU list */
    list_add_after(&c->lists[hash], &elem->link);
}

static inline void
cache_item_set_unused(listcache *c, cache_link *elem)
{
    list_add_before(&c->lru_list, &elem->lru_link);
    c->lru_entries++;
}

static inline cache_link *
cache_find(listcache *c, unsigned int val)
{
    unsigned int hash = val % CACHE_LISTS;
    list_elem *elem;

    list_for_each_elem(&c->lists[hash], elem) {
        cache_link *l = container_of(elem, cache_link, link);
        if (c->elem_val(l) == val) {
            if (!list_empty(&l->lru_link)) {
                /* It's in the unused list, remove it. */
                list_del(&l->lru_link);
                list_item_init(&l->lru_link);
                c->lru_entries--;
            }
            return l;
        }
    }
    return NULL;
}

static inline int
cache_flush(listcache *c)
{
    list_elem *elem, *next;
    cache_link *l;
    int i;

    list_for_each_elem_safe(&c->lru_list, elem, next) {
        l = container_of(elem, cache_link, lru_link);
        list_del(elem);
        list_del(&l->link);
        c->entries--;
        c->lru_entries--;
        c->freed(l);
    }

    for (i = 0; i < CACHE_LISTS; i++) {
        list_for_each_elem_safe(&c->lists[i], elem, next) {
            l = container_of(elem, cache_link, link);
            list_del(&l->link);
            c->entries--;
            c->freed(l);
        }
    }

    return c->entries || c->lru_entries;
}

static inline void
cache_init(listcache *c, unsigned int max_free_entries,
       unsigned int (*elem_val)(cache_link *elem),
       void (*freed)(cache_link *elem))
{
    int i;

    c->entries = 0;
    c->lru_entries = 0;
    c->max_free_entries = max_free_entries;
    list_init(&c->lru_list);
    for (i = 0; i < CACHE_LISTS; i++)
        list_init(&c->lists[i]);
    c->elem_val = elem_val;
    c->freed = freed;
}

#endif /* __CACHE_H__ */


test-mount.sh
#!/bin/sh

# Use this script if you need to regenerate the digest values
# in test.sh, or if you don't care about digests and you just
# want to see some fsck results. Should be run as root.

# Each test here creates a test image, verifies it and prints
# the command line for use in test.sh for regression testing.

set -e

. ./test-gen.lib

test_mnt=t_mnt

test_common () {
	/sbin/e2fsck -fn $test_img || fail
	mkdir $test_mnt
	mount -t ext2 -o ro,loop $test_img $test_mnt || fail
}

test_cleanup () {
	umount $test_mnt
	rmdir $test_mnt
	rm -f fout lsout
}

fail () {
	echo FAIL
	test_cleanup
	gen_cleanup
	exit 1
}

pass () {
	cmd=$1
	shift
	md5=`calc_digest`
	echo PASS
	echo $cmd $md5 $@
	test_cleanup
	gen_cleanup
}

# dtest_mount - Exercise the -d option of genext2fs.
dtest_mount () {
	size=$3
	dgen $@
	test_common
	test -f $test_mnt/file.$size || fail
	test $size = "`ls -al $test_mnt | \
	               grep file.$size | \
	               awk '{print $5}'`" || fail
	pass dtest $@
}

# ftest_mount - Exercise the -D option of genext2fs.
ftest_mount () {
	fname=$3
	fgen $@
	test_common
	test -d $test_mnt/dev || fail
	# Exclude those devices that have interpolated
	# minor numbers, as being too hard to match.
	egrep -v "(hda|hdb|tty|loop|ram|ubda)" $fname | \
		grep '^[^	#]*	[bc]' | \
		awk '{print $1,$4,$5,$6","$7}'| \
		sort -d -k3.6 > fout
	ls -aln $test_mnt/dev | \
		egrep -v "(hda|hdb|tty|loop|ram|ubda)" | \
		grep ^[bc] | \
		awk '{ print "/dev/"$10,$3,$4,$5$6}' | \
		sort -d -k3.6 > lsout
	diff fout lsout || fail
	pass ftest $@
}

# ltest_mount - Exercise the -d option of genext2fs, with symlink.
ltest_mount () {
	appendage=$3
	lgen $@
	test_common
	cd $test_mnt
	readlink symlink > ../lsout
	cd ..
	test -s lsout || fail
	echo 12345678901234567890123456789012345678901234567890$appendage > fout
	diff fout lsout || fail
	pass ltest $@
}

dtest_mount 4096 1024 0
dtest_mount 2048 2048 0
dtest_mount 1024 4096 0
dtest_mount 8193 1024 0
dtest_mount 8194 1024 0
dtest_mount 8193 4096 0
dtest_mount 8194 2048 0
dtest_mount 4096 1024 1
dtest_mount 1024 4096 1
dtest_mount 4096 1024 12288
dtest_mount 4096 1024 274432
dtest_mount 9000 1024 8388608
dtest_mount 4500 2048 8388608
dtest_mount 2250 4096 8388608
dtest_mount 20000 1024 16777216
dtest_mount 10000 2048 16777216
ftest_mount 4096 default device_table.txt
ltest_mount 200 1024 123456789
ltest_mount 200 1024 1234567890
ltest_mount 200 4096 12345678901
